File: techniques/homogeneity-of-variance/r/homogeneity_of_variance.R
# Tests for homogeneity of variance across k groups (Reference §3.20, §3.55)
# From-scratch base R plus idiomatic calls (car::leveneTest, stats::bartlett.test,
# stats::var.test).  Run with:  Rscript homogeneity_of_variance.R
#
# Inputs used below:
#   groups : list of numeric vectors (one per group)
#   center : "mean" or "median" (Levene vs. Brown-Forsythe)
#   x1, x2 : two-sample inputs for the F-test of variances

levene_bf <- function(groups, center = "median") {
  k <- length(groups); ns <- vapply(groups, length, integer(1))
  ctr <- vapply(groups, if (center == "median") median else mean, numeric(1))
  z <- mapply(function(g, c0) abs(g - c0), groups, ctr, SIMPLIFY = FALSE)
  N <- sum(ns); grand_z <- mean(unlist(z))
  z_means <- vapply(z, mean, numeric(1))
  ss_b <- sum(ns * (z_means - grand_z)^2)
  ss_w <- sum(unlist(mapply(function(zi, m) (zi - m)^2, z, z_means, SIMPLIFY = FALSE)))
  df1 <- k - 1; df2 <- N - k
  F <- (ss_b / df1) / (ss_w / df2)
  list(statistic = F, df1 = df1, df2 = df2,
       p_value = pf(F, df1, df2, lower.tail = FALSE), center = center)
}

levene_scratch <- function(groups) levene_bf(groups, center = "mean")
brown_forsythe_scratch <- function(groups) levene_bf(groups, center = "median")

bartlett_scratch <- function(groups) {
  k <- length(groups); ns <- vapply(groups, length, integer(1))
  vars <- vapply(groups, var, numeric(1))
  N <- sum(ns); sp2 <- sum((ns - 1) * vars) / (N - k)
  num <- (N - k) * log(sp2) - sum((ns - 1) * log(vars))
  denom <- 1 + (1 / (3 * (k - 1))) * (sum(1 / (ns - 1)) - 1 / (N - k))
  chi2 <- num / denom
  list(statistic = chi2, df = k - 1,
       p_value = pchisq(chi2, k - 1, lower.tail = FALSE))
}

f_test_two_variances_scratch <- function(x1, x2, alternative = "two.sided") {
  v1 <- var(x1); v2 <- var(x2); F <- v1 / v2
  df1 <- length(x1) - 1; df2 <- length(x2) - 1
  p <- switch(alternative,
    "two.sided" = 2 * min(pf(F, df1, df2), pf(F, df1, df2, lower.tail = FALSE)),
    "greater"   = pf(F, df1, df2, lower.tail = FALSE),
    "less"      = pf(F, df1, df2))
  list(F = F, df1 = df1, df2 = df2, p_value = min(1, p), var1 = v1, var2 = v2)
}

# Library: car::leveneTest (with center =), stats::bartlett.test, stats::var.test
library_demo <- function(groups, x1, x2) {
  values <- unlist(groups)
  grp <- factor(rep(seq_along(groups), sapply(groups, length)))
  if (requireNamespace("car", quietly = TRUE)) {
    cat("car::leveneTest (center = median, default):\n")
    print(car::leveneTest(values ~ grp))
  } else cat("(install 'car' for leveneTest)\n")
  cat("\nstats::bartlett.test:\n");  print(bartlett.test(values, grp))
  cat("\nstats::var.test (a vs c):\n"); print(var.test(x1, x2))
}

if (sys.nframe() == 0) {
  set.seed(4)
  a <- rnorm(30, 50, 10); b <- rnorm(28, 50, 11); c <- rnorm(32, 50, 22)
  cat("=== Levene (center = mean) ===\n");        print(levene_scratch(list(a,b,c)))
  cat("\n=== Brown-Forsythe (center = median) ===\n"); print(brown_forsythe_scratch(list(a,b,c)))
  cat("\n=== Bartlett ===\n");                    print(bartlett_scratch(list(a,b,c)))
  cat("\n=== F-test of variances (a vs c) ===\n"); print(f_test_two_variances_scratch(a, c))
  cat("\n--- library ---\n");                     library_demo(list(a,b,c), a, c)
}

File: techniques/homogeneity-of-variance/python/homogeneity_of_variance.py
"""Tests for homogeneity of variance across k groups (Reference §3.20, §3.55).

These check the ANOVA / Student's-t assumption that all groups have equal
variances. Three classical tests, in decreasing order of sensitivity to
non-normality:

  - Levene's test      : one-way ANOVA on |x_ij - mean(x_i)|   (center = mean)
  - Brown-Forsythe     : one-way ANOVA on |x_ij - median(x_i)| (center = median)
                         More robust to non-normality than classic Levene.
  - Bartlett's test    : likelihood-ratio against equal variances. Most powerful
                         when groups ARE normal; very sensitive to non-normality
                         otherwise -- don't use unless you've checked normality.
  - F-test of variances: for the two-sample case (s1^2 / s2^2 ~ F(n1-1, n2-1)).
                         Same non-normality caveat as Bartlett.

If any test is significant -> use Welch's ANOVA / Welch's t / Games-Howell
post-hoc instead of the equal-variance versions.
"""
from __future__ import annotations    # stdlib: postpone type-hint evaluation (lets us write int | None)

import math    # stdlib: scalar math (sqrt, log, exp, comb, lgamma, pi, ...)
from typing import Sequence    # stdlib: type hint meaning 'indexable iterable' (list / tuple / array)

from scipy import stats    # distributions, hypothesis tests, PPFs (norm, t, chi2, ttest_ind, ...)


def _mean(g): return sum(g) / len(g)


def _median(g):
    s = sorted(g); n = len(s)
    return s[n // 2] if n % 2 else (s[n // 2 - 1] + s[n // 2]) / 2.0


def _levene_brown_forsythe(groups, center="median"):
    """Levene's test with configurable center.

    center = "mean"   -> Levene's original test
    center = "median" -> Brown-Forsythe (more robust)
    """
    k = len(groups)
    ns = [len(g) for g in groups]
    centers = [(_mean(g) if center == "mean" else _median(g)) for g in groups]
    # within-group absolute deviations
    z = [[abs(v - c) for v in g] for g, c in zip(groups, centers)]
    N = sum(ns)
    grand_z = sum(sum(zi) for zi in z) / N
    z_means = [sum(zi) / n for zi, n in zip(z, ns)]
    ss_b = sum(n * (zm - grand_z) ** 2 for n, zm in zip(ns, z_means))
    ss_w = sum((v - zm) ** 2 for zi, zm in zip(z, z_means) for v in zi)
    df1, df2 = k - 1, N - k
    if ss_w == 0:
        return {"statistic": float("inf"), "df1": df1, "df2": df2, "p_value": 0.0,
                "center": center}
    F = (ss_b / df1) / (ss_w / df2)
    p = float(stats.f.sf(F, df1, df2))
    return {"statistic": F, "df1": df1, "df2": df2, "p_value": p, "center": center}


def levene(groups: Sequence[Sequence[float]]) -> dict:
    """Levene's test (center = mean)."""
    return _levene_brown_forsythe(groups, center="mean")


def brown_forsythe(groups: Sequence[Sequence[float]]) -> dict:
    """Brown-Forsythe variance test (Levene with center = median). Robust default."""
    return _levene_brown_forsythe(groups, center="median")


def bartlett(groups: Sequence[Sequence[float]]) -> dict:
    """Bartlett's test (likelihood-ratio). High power IF groups are normal."""
    k = len(groups)
    ns = [len(g) for g in groups]
    vars_ = [sum((v - _mean(g)) ** 2 for v in g) / (n - 1) for g, n in zip(groups, ns)]
    N = sum(ns)
    sp2 = sum((n - 1) * v for n, v in zip(ns, vars_)) / (N - k)
    num = (N - k) * math.log(sp2) - sum((n - 1) * math.log(v) for n, v in zip(ns, vars_))
    denom = 1 + (1 / (3 * (k - 1))) * (
        sum(1 / (n - 1) for n in ns) - 1 / (N - k)
    )
    chi2 = num / denom
    p = float(stats.chi2.sf(chi2, k - 1))
    return {"statistic": chi2, "df": k - 1, "p_value": p}


def f_test_two_variances(x1: Sequence[float], x2: Sequence[float],
                         alternative: str = "two-sided") -> dict:
    """F-test of H0: var(x1) == var(x2). Tiny power loss vs. Bartlett for k=2."""
    n1, n2 = len(x1), len(x2)
    v1 = sum((v - _mean(x1)) ** 2 for v in x1) / (n1 - 1)
    v2 = sum((v - _mean(x2)) ** 2 for v in x2) / (n2 - 1)
    F = v1 / v2
    df1, df2 = n1 - 1, n2 - 1
    if alternative == "two-sided":
        p = 2 * min(stats.f.cdf(F, df1, df2), stats.f.sf(F, df1, df2))
    elif alternative == "greater":
        p = float(stats.f.sf(F, df1, df2))
    elif alternative == "less":
        p = float(stats.f.cdf(F, df1, df2))
    else:
        raise ValueError("alternative must be 'two-sided', 'greater', or 'less'")
    return {"F": F, "df1": df1, "df2": df2, "p_value": min(1.0, p), "var1": v1, "var2": v2}


def library_versions(groups):
    return {
        "scipy.stats.levene (median)": stats.levene(*groups, center="median"),
        "scipy.stats.levene (mean)":   stats.levene(*groups, center="mean"),
        "scipy.stats.bartlett":         stats.bartlett(*groups),
    }


if __name__ == "__main__":
    import numpy as np
    rng = np.random.default_rng(4)
    # group C has noticeably larger variance
    a = rng.normal(50, 10, 30).tolist()
    b = rng.normal(50, 11, 28).tolist()
    c = rng.normal(50, 22, 32).tolist()
    groups = [a, b, c]

    print("=== Levene's test (center=mean) ===")
    print(levene(groups))
    print("\n=== Brown-Forsythe (center=median) ===")
    print(brown_forsythe(groups))
    print("\n=== Bartlett's test ===")
    print(bartlett(groups))
    print("\n=== F-test (two-sample): a vs c ===")
    print(f_test_two_variances(a, c))

    print("\n--- library (scipy) ---")
    for k, v in library_versions(groups).items():
        print(f"  {k}: {v}")
